docs/best-practice-consensus-pipeline.md

```diff
@@ -0,0 +1,217 @@
+
+```bash
+fgbio -Xmx1g --compression 1 --async-io FastqToBam \
+  --input r1.fq.gz r2.fq.gz \
+  --read-structures 8M+T +T \
+  --sample s1 \
+  --library library1 \
+  --platform-unit ABCDEAAXX.1 \
+  --output s1.unmapped.bam
+```
+
+If, however, the UMIs are a limited and non-fully degenerate set then they can be error-corrected after extraction from the reads as follows:
+  
+```bash
+fgbio -Xmx1g --compression 0 --async-io FastqToBam \
+  --input r1.fq.gz r2.fq.gz \
+  --read-structures 8M+T +T \
+  --sample s1 \
+  --library library1 \
+  --platform-unit ABCDEAAXX.1 \
+  --output /dev/stdout \
+  | fgbio -Xmx2g --compression 1 CorrectUmis \
+      --input /dev/stdin \
+      --output s1.unmapped.bam \
+      --max-mismatches 1 \
+      --min-distance 2 \
+      --umi-files file-of-expected-umis-one-per-line.txt \
+      --dont-store-original-umis
+```
+
+Neither option above requires or performs any sorting.
+
+### Step 1.2: uBam -> Mapped BAM
+
+This step uses a combination of `samtools` (to convert the uBam back to FASTQ), `bwa mem` to do the actual alignment and `fgbio ZipperBams` to carry metaadata and UMIs forward from the unmapped BAM.  No sorting is performed in this step.
+
+```  
+# Align the data with bwa and recover headers and tags from the unmapped BAM
+samtools fastq s1.unmapped.bam \
+  | bwa mem -t 16 -p -K 150000000 -Y ref.fa \
+  | fgbio -Xmx4g --compression 1 --async-io ZipperBams \
+      --unmapped-bam s1.unmapped.bam \
+      --ref ref.fa \
+      --output s1.mapped.bam 
+```
+
+Some important `bwa mem` options used here are:
+
+- `-t 16` to set the number of threads to use to 16; this can be set to any number between 1 and the number of CPU cores available
+- `-K 150000000` to tell `bwa` to process reads/bases in fixed size chunks regardless of the number of threads; this makes `bwa` deterministic and reproducible with different thread counts
+- `-Y` to use soft-clipping (instead of hard clipping) in supplemntal alignments
+
+Under no circumstances should `-M` be used to `mark shorter split hits as secondary` as this will cause problems with downstream tools.
+
+
+### Step 1.3: Mapped BAM -> Grouped BAM
+
+This step identifies reads or read pairs that originate from the same source molecule based on genomic positions and UMI.  For _most_ data this is the recommended way to run grouping:
+
+```  
+# Group the reads by position and UMI
+fgbio -Xmx8g --compression 1 GroupReadsByUmi \
+  --input s1.mapped.bam \
+  --strategy Adjacency \
+  --edits 1 \
+  --output s1.grouped.bam \
+  --family-size-histogram s1.tag-family-sizes.txt
+```
+
+However:
+
+1. If UMIs are significantly longer (e.g. 20bp) or have more errors, `--edits` can be increased
+2. For multiplex PCR and similar data where reads' genomic positions are fixed by the primers it is recommended to use `--strategy Identity` to reduce runtime at the expense of lower accuracy
+
```

Comment:
The mapped BAM may be streamed directly into `GroupReadsByUmi`.

Reply:
I am curious as to why not just stream into `GroupReadsByUmi`, since the latter is going to have to sort anyways.  Or stream the mapped reads into `fgbio SortSam` (ugh)

Reply:
My 2c is that since `GroupReadsByUmi` discards reads, I like to save the aligned/zippered BAM before pushing it through GroupReadsByUmi.  That way if I want to change my mind about --min-map-q or other settings that affect which reads to use, I don't have to re-align the data.
Also, frankly, a lot of people will branch off at this point to run metrics, alt analysis on non-consensus data, etc.